### PR TITLE
#10370 fix

### LINKED ISSUE #10370: [bug] (added) inventory item title is always first listed modules name
# 🐞 bug report

### ⁉️ Describe the bug
If I add new different inventory elements, the element title is always the same

### 🔥 How to trigger the error
bug_0010

### 👎 Actual Behavior
new element title is always the same

### 👍 Expected Behavior
each element with correct title

### 📷 Screenshot of configuration
[none]

### 📝 PHP/Apache/Nginx/Browser/CRM Logs
[indipendent]

### 🌍  Your Environment 
[indipendent]

### ❓ Additional context
[none]

### :heavy_check_mark: Fix
Put translated "SINGLE_" strings (json format) into (new) data-inventorylbls attribute for tbody.js-inventory-base-item (6 SMARTY-code lines).
Get correct label by JS (2 JS-code lines).

in /var/www/yetiforce/layouts/basic/modules/Vtiger/Edit/**Inventory.tpl:154** add following lines
```code
{assign var="INVENTORYLBLS" value=[]}
{foreach item=MAIN_MODULE from=$MAIN_PARAMS['modules']}
	{assign var=STR value="SINGLE_"|cat:$MAIN_MODULE}
	{$INVENTORYLBLS[$STR]=\App\Language::translate($STR, $MAIN_MODULE)}
{/foreach}
<tbody class="js-inventory-base-item" data-inventorylbls="{App\Purifier::encodeHtml(\App\Json::encode($INVENTORYLBLS))}">
```

full code for **Inventory.tpl**
```code
{include file=\App\Layout::getTemplatePath('Edit/InventorySummary.tpl', $MODULE_NAME)}
{assign var="ITEM_DATA" value=$RECORD->getInventoryDefaultDataFields()}
<table id="blackIthemTable" class="noValidate d-none">
	
	
	{assign var="INVENTORYLBLS" value=[]}
	{foreach item=MAIN_MODULE from=$MAIN_PARAMS['modules']}
		{assign var=STR value="SINGLE_"|cat:$MAIN_MODULE}
		{$INVENTORYLBLS[$STR]=\App\Language::translate($STR, $MAIN_MODULE)}
	{/foreach}
	<tbody class="js-inventory-base-item" data-inventorylbls="{App\Purifier::encodeHtml(\App\Json::encode($INVENTORYLBLS))}">
	
	
	{assign var="ROW_NO" value='_NUM_'}
	{include file=\App\Layout::getTemplatePath('Edit/InventoryItem.tpl', $MODULE_NAME)}
	</tbody>
</table>
```

in /public_html/layouts/basic/modules/Vtiger/resources/Inventory.min.js into **function addItem(module, baseTableId)**
add following lines

```code
...
inventorylbls = newRow.data('inventorylbls')
...
newRow.find('.rowName span.input-group-text').attr('data-content', inventorylbls['SINGLE_' + module])
...
```

full code for **function addItem(module, baseTableId)**
```code
addItem: function addItem(module, baseTableId) {
	var rowData = !!(2 < arguments.length && arguments[2] !== void 0) && arguments[2]
		, items = this.getInventoryItemsContainer()
		, newRow = this.getBasicRow()
		, sequenceNumber = this.getNextLineItemRowNumber()
		, replaced = newRow.html().replace(/\_NUM_/g, sequenceNumber)
		, inventorylbls = newRow.data('inventorylbls'); // <<<<< added this line ******************
	newRow.html(replaced),
	newRow = newRow.children().appendTo(items.find('.js-inventory-items-body')),
	newRow.find('.rowName input[name="popupReferenceModule"]').val(module).data('field', baseTableId),
	newRow.find('.js-module-icon').removeClass().addClass('userIcon-' + module),
	newRow.find('.rowName span.input-group-text').attr('data-content', inventorylbls['SINGLE_' + module]), // <<<<< added this line ******************
	newRow.find('.colPicklistField select').each(function(index, select) {
		select = $(select),
		select.find('option').each(function(index, option) {
			option = $(option),
			option.data('module') !== module && option.remove();
		});
	}),
	this.initItem(newRow),
	Vtiger_Edit_Js.getInstance().registerAutoCompleteFields(newRow),
	rowData && this.setRowData(newRow, rowData);
},
```

### PR DESCRIPTION
**Pull request by davide-alghi**
## Description
Put translated "SINGLE_" strings (json format) into (new) data-inventorylbls attribute for tbody.js-inventory-base-item (Inventory.tpl)
Get correct label by JS (Inventory.min.js)

fix #10370 issue

## Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by davide-alghi** [comment]
If I can, I have some technical questions.

- js files are in both beautified and minified format: why are both files loaded by YF?
- Is it possible there are differences between beautified and minified files? About class methods and / or functions (I'm not sure, but sometimes I have had to modify minified file because part of code was not in beautified file - this is the reason why I modified min.js file in this my pull request)
- about css files, it seems that only beautified files are loaded by YF
- how and where have js and css files to be modified?

**Comment by mariuszkrzaczkowski** [comment]
> public_html/layouts/basic/modules/Vtiger/resources/Inventory.min.js

In min files, we do not make changes only in the full file,
min files are generated automatically

**Comment by davide-alghi** [comment]
Thank Mariusz,
I'm going to change to public_html/layouts/basic/modules/Vtiger/resources/Inventory.min.js the fix


about 1st point, I was confused by FF Debugger: only the min.js files are loaded
for 2nd, you replied me,
3rd and 4th points I'll open an issue

**Comment by mariuszkrzaczkowski** [comment]
disable loading of min files, it will be easier
config/developer.php 

```php
/** Enable minimize JS files */
public static $MINIMIZE_JS = false;
```

https://github.com/YetiForceCompany/YetiForceCRM/blob/developer/config/ConfigTemplates.php#L446

**Comment by davide-alghi** [comment]
yep ;)

**Comment by mariuszkrzaczkowski** [comment]
> js files are in both beautified and minified format: why are both files loaded by YF?

What do you mean? Always one min or full file is loaded
> Is it possible there are differences between beautified and minified files? About class methods and / or functions 
> (I'm not sure, but sometimes I have had to modify minified file because part of code was not in beautified file - this is the reason why I modified min.js file in this my pull request)

yes it's possible because we build min files once in a while, not each time the file changes, same applies to CSS files. If you have to modify a JS file it's best to disable MIN files
> about css files, it seems that only beautified files are loaded by YF

by default yes but it can be changed
> how and where have js and css files to be modified?

we always modify only full files, all files are always built before a version is published, and on the dev version it happens once in a while

**Comment by davide-alghi** [comment]
> > js files are in both beautified and minified format: why are both files loaded by YF?
> 
> What do you mean? Always one min or full file is loaded
> 
Please forget that, I was confused by FF Debugger: it lists both files (if $MINIMIZE_JS is true), maybe because *.js.map :(
> > Is it possible there are differences between beautified and minified files? About class methods and / or functions
> > (I'm not sure, but sometimes I have had to modify minified file because part of code was not in beautified file - this is the reason why I modified min.js file in this my pull request)
> 
> yes it's possible because we build min files once in a while, not each time the file changes, same applies to CSS files. If you have to modify a JS file it's best to disable MIN files
> 
ok, I got it
> > about css files, it seems that only beautified files are loaded by YF
> 
> by default yes but it can be changed
> config/developer.php
> 
> ```
> /** Enable minimize CSS files */
> public static $MINIMIZE_CSS = false;
> ```
> 
this happens if YF integrator doesn't point the webserver document-root to YF/public_html (inner one). Assuming both $MINIMIZE_JS and $MINIMIZE_CSS to "true", and the webserver document-root **doesn't point** to YF/public_html, js files are loaded in minified version, instead css file are (always) loaded in full version. This because **public function checkAndConvertJsScripts($jsFileNames){...}** and **public function checkAndConvertCssStyles($cssFileNames, $fileExtension = 'css') {...}** have just a little different mode to decide if load min or full files
> > how and where have js and css files to be modified?
> 
> we always modify only full files, all files are always built before a version is published, and on the dev version it happens once in a while

now I got it: nextly I'll debug and modify only js/css beauty version.
it would be useful to know with which tools YF-team minimizes css / js files to test even if they are minimized correctly, before proposing them. I assume that limit is css3 and ES5 for complete cross-browser compatibility

**Comment by mariuszkrzaczkowski** [comment]
> > > js files are in both beautified and minified format: why are both files loaded by YF?
> > 
> > 
> > What do you mean? Always one min or full file is loaded
> 
> Please forget that, I was confused by FF Debugger: it lists both files (if $MINIMIZE_JS is true), maybe because *.js.map :(
> 
> > > Is it possible there are differences between beautified and minified files? About class methods and / or functions
> > > (I'm not sure, but sometimes I have had to modify minified file because part of code was not in beautified file - this is the reason why I modified min.js file in this my pull request)
> > 
> > 
> > yes it's possible because we build min files once in a while, not each time the file changes, same applies to CSS files. If you have to modify a JS file it's best to disable MIN files
> 
> ok, I got it
> 
> > > about css files, it seems that only beautified files are loaded by YF
> > 
> > 
> > by default yes but it can be changed
> > config/developer.php
> > ```
> > /** Enable minimize CSS files */
> > public static $MINIMIZE_CSS = false;
> > ```
> 
> this happens if YF integrator doesn't point the webserver document-root to YF/public_html (inner one). Assuming both $MINIMIZE_JS and $MINIMIZE_CSS to "true", and the webserver document-root **doesn't point** to YF/public_html, js files are loaded in minified version, instead css file are (always) loaded in full version. This because **public function checkAndConvertJsScripts($jsFileNames){...}** and **public function checkAndConvertCssStyles($cssFileNames, $fileExtension = 'css') {...}** have just a little different mode to decide if load min or full files

add a issue we will check it

> 
> > > how and where have js and css files to be modified?
> > 
> > 
> > we always modify only full files, all files are always built before a version is published, and on the dev version it happens once in a while
> 
> now I got it: nextly I'll debug and modify only js/css beauty version.
> it would be useful to know with which tools YF-team minimizes css / js files to test even if they are minimized correctly, before proposing them. I assume that limit is css3 and ES5 for complete cross-browser compatibility

we've published the code for compiling files
https://github.com/YetiForceCompany/YetiForceScripts/tree/master/YetiDevsTools

File: layouts/basic/modules/Vtiger/Edit/Inventory.tpl [hunk 1]
```diff
@@ -152,7 +152,12 @@
 		{include file=\App\Layout::getTemplatePath('Edit/InventorySummary.tpl', $MODULE_NAME)}
 		{assign var="ITEM_DATA" value=$RECORD->getInventoryDefaultDataFields()}
 		<table id="blackIthemTable" class="noValidate d-none">
-			<tbody class="js-inventory-base-item">
+			{assign var="INVENTORYLBLS" value=[]}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/Edit/Inventory.tpl]:
```suggestion
			{assign var="INVENTORY_LBLS" value=[]}
```

File: layouts/basic/modules/Vtiger/Edit/Inventory.tpl [hunk 2]
```diff
@@ -152,7 +152,12 @@
...
+			{foreach item=MAIN_MODULE from=$MAIN_PARAMS['modules']}
+				{assign var=STR value="SINGLE_"|cat:$MAIN_MODULE}
+				{$INVENTORYLBLS[$STR]=\App\Language::translate($STR, $MAIN_MODULE)}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/Edit/Inventory.tpl]:
```suggestion
				{$INVENTORY_LBLS[$MAIN_MODULE]=\App\Language::translateSingularModuleName($MAIN_MODULE)}
```

File: layouts/basic/modules/Vtiger/Edit/Inventory.tpl [hunk 3]
```diff
@@ -152,7 +152,12 @@
 		{include file=\App\Layout::getTemplatePath('Edit/InventorySummary.tpl', $MODULE_NAME)}
 		{assign var="ITEM_DATA" value=$RECORD->getInventoryDefaultDataFields()}
 		<table id="blackIthemTable" class="noValidate d-none">
-			<tbody class="js-inventory-base-item">
+			{assign var="INVENTORYLBLS" value=[]}
+			{foreach item=MAIN_MODULE from=$MAIN_PARAMS['modules']}
+				{assign var=STR value="SINGLE_"|cat:$MAIN_MODULE}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/Edit/Inventory.tpl]: to remove unnecessary

File: layouts/basic/modules/Vtiger/Edit/Inventory.tpl [hunk 4]
```diff
@@ -152,7 +152,12 @@
...
+				{$INVENTORYLBLS[$STR]=\App\Language::translate($STR, $MAIN_MODULE)}
+			{/foreach}
+			<tbody class="js-inventory-base-item" data-inventorylbls="{App\Purifier::encodeHtml(\App\Json::encode($INVENTORYLBLS))}">
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/Edit/Inventory.tpl]:
```suggestion
			<tbody class="js-inventory-base-item" data-module-lbls="{App\Purifier::encodeHtml(\App\Json::encode($INVENTORY_LBLS))}">
```

File: public_html/layouts/basic/modules/Vtiger/resources/Inventory.js [hunk 5]
```diff
@@ -1333,6 +1333,7 @@ $.Class(
 			let newRow = this.getBasicRow();
 			const sequenceNumber = this.getNextLineItemRowNumber();
 			const replaced = newRow.html().replace(/\_NUM_/g, sequenceNumber);
+			const inventorylbls = newRow.data('inventorylbls');
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Inventory.js]:
```suggestion
			const moduleLbls = newRow.data('moduleLbls');
```

File: public_html/layouts/basic/modules/Vtiger/resources/Inventory.js [hunk 6]
```diff
@@ -1343,6 +1344,9 @@ $.Class(
 				.find('.js-module-icon')
 				.removeClass()
 				.addClass(`userIcon-${module}`);
+			newRow
+				.find('.rowName span.input-group-text')
+				.attr('data-content', inventorylbls['SINGLE_' + module]);
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/Inventory.js]:
```suggestion
				.attr('data-content', moduleLbls[module]);
```

File: layouts/basic/modules/Vtiger/Edit/Inventory.tpl (hunk 1 above)
Comment (davide-alghi) [reply layouts/basic/modules/Vtiger/Edit/Inventory.tpl]: yes, they're better

**Comment by davide-alghi** [comment]
Hi Mariusz,
good your suggestions.

just two questions
in "Files changed", have I to **Commit suggestion** or you have to do that?
what's the difference with **Add suggestion to batch**?

**Comment by mariuszkrzaczkowski** [comment]
If the suggestions don't require any other changes then they can be accepted and a change will be automatically added.
In your case you can upload the changes.

**Pull request merged**